Improve Attributes Handling

gregw opened the pull request:
Spun out from #4814 

Improve attribute handling to reduce garbage and improve lookup.
Introduced a Wrapper so that request can remove any layers on reset.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Spun out from #4814 Improve Attributes Handling

Improve attribute handling to reduce garbage and improve lookup.
Introduced a Wrapper so that request can remove any layers on reset.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:
```diff
@@ -353,36 +317,67 @@ public String toString()
+            }
+            catch (Exception e)
+            {
+                LOG.warn("Unable to customize request with encryption details", e);
```
This is a copy/paste?  The message isn't valid for this new wrapper version.

joakime commented on jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:
```diff
@@ -353,36 +317,67 @@ public String toString()
+            for (Enumeration<String> e = _attributes.getAttributeNames(); e.hasMoreElements(); )
+            {
+                names.add(e.nextElement());
+            }
```
This could be ...

``` java
Set<String> names = new HashSet<>();
_attributes.getAttributeNames().asIterator().forEachRemaining(names::add);
```

I'm just mentioning this here to show a different way, not that it's a better way (it likely isn't).

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1867,8 +1867,14 @@ protected void recycle()
         if (_attributes != null)
-            _attributes.clearAttributes();
+        {
+            if (AttributesMap.class.equals(_attributes.getClass()))
```
Why cant you use `instance of` here instead of `class.equals`?

gregw replied:
Because classes that extend AttributesMap with unknown behaviours will pass an instanceof test.
I  already know it is an instanceof `AttributesMap` because that is the type of the field.  I want to ensure it is a base implementation that we can clear and not have any special behaviour left over.

joakime committed:
Issue #4814 - Exposing AttributeMap.getAttributeNameSet() on Attributes.

The underlying AttributesMap already has a .getAttributeNameSet()
method, expose it on the Attributes interface.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime reviewed (commented):
I added commit d460786b94f06bca5c363b52ed9cf080db30a4aa to this PR to expose the AttributesMap.getAttributeNameSet() on the Attributes interface and use it.

Feel free to reject / rollback it if you don't like that change.

Since this PR is an effort to reduce GC, this change eliminates the intermediary Enumeration on some code paths.

lachlan-roberts commented on jetty-server/src/main/java/org/eclipse/jetty/server/SecureRequestCustomizer.java:
```diff
@@ -353,36 +317,67 @@ public String toString()
+        private final SSLSession _session;
 
-        X509Certificate[] getCerts()
+        public SslAttributes(Request request, SSLSession sslSession, Attributes attributes)
```
This class now has a weird behaviour that if you set one of these `JAVAX_SERVLET_REQUEST_*` attributes, it sets it on the wrapped attributes, but when accessing the attributes you only get the version stored here. Maybe you should override `setAttribute` as well.

gregw replied:
yeah probably.  Just not clear what it should do?  Replace the passed value?
Maybe in get, we should try the wrapped one first and then only use the special handling if it is null.   Yeah will do it that way....

lachlan-roberts replied:
So what happens now if an `SslAttributes` was wrapping another `Attributes` with one of these attributes already set, you would want `SslAttributes` to override the behaviour of the wrapped attributes and not the other way around.

I feel like it would be a better fix to store the SSL attributes at this level only, similar to the pattern used in `ForwardAttributes` and `IncludeAttributes`. Seems weird that they would be stored both here and in the wrapped attributes.

But I'll just approve and leave it up to you, PR looks good to me in general.

gregw commented:
@joakime I like the Set access, but that method now also needs to be overridden in the wrapping Attributes... actually if done right then `getAttributeNameSet` needs to be overridden and `getAttributeNames` can use that.  As it currently is, any body calling `getAttributeNameSet` will not see the extra names on a forward or include.  Can you fix?

joakime commented:
> @joakime I like the Set access, but that method now also needs to be overridden in the wrapping Attributes... actually if done right then `getAttributeNameSet` needs to be overridden and `getAttributeNames` can use that. As it currently is, any body calling `getAttributeNameSet` will not see the extra names on a forward or include. Can you fix?

Yup, I can take care of that.

gregw committed:
Allow a set to override a secure attribute.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime committed:
Issue #4814 - Attributes.getAttributeNames() is now defaulted

The Attributes.getAttributeNames() will use the
.getAttributeNameSet() by default now.

Updated all Attributes.Wrapper impls to use this new behavior

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

joakime committed:
Merge branch 'jetty-9.4.x-4814-ImproveAttributeHandling' of github.com:eclipse/jetty.project into jetty-9.4.x-4814-ImproveAttributeHandling

joakime commented:
@gregw the change you asked for to impl Set in wrapper can be found in commit 663ef83003039431e876d1f5071f73e9cab0dc94

gregw commented:
I think no matter what, if one of those attributes is set we should return that value.
